### PR TITLE
Forward custom header names

### PR DESCRIPTION
**Pull request by moander**
Made it possible to use custom header names for X-Forwarded-For|Host|Proto

### CODE REVIEW

**Comment by Tratcher** [comment]
What other names did you plan to use? Are you sure the semantics of those other headers are the same?

**Comment by moander** [comment]
The developer should be allowed to decide what other names to use, and also be responsible for making sure the semantics of those are the same. I agree that my examples are irresponsible as they have not been tested properly.

**Comment by davidfowl** [comment]
> The developer should be allowed to decide what other names to use, and also be responsible for making sure the semantics of those are the same. I agree that my examples are irresponsible as they have not been tested properly.

Why? What real scenario inspired this change in the first place?

**Comment by moander** [comment]
Google App Engine

**Comment by davidfowl** [comment]
What names do they use?

**Comment by moander** [comment]
X-Google-Real-IP

Try to search for ` x-forwarded-for` on Google if you want more examples. There are plenty :)

Here is one example found in the HAProxy docs:

The keyword "header" may be used to supply a different header name to replace the default "X-Forwarded-For". This can be useful where you might already have a "X-Forwarded-For" header from a different application (eg: stunnel), and you need preserve it. Also if your backend server doesn't use the "X-Forwarded-For" header and requires different one (eg: Zeus Web Servers require "X-Cluster-Client-IP").

**Comment by davidfowl** [comment]
I don't have any pushback against the feature. It just needs to exist for good reasons.

File: src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs [hunk 1]
```diff
@@ -17,9 +17,18 @@ namespace Microsoft.AspNetCore.HttpOverrides
...
-        private const string XForwardedProtoHeaderName = "X-Forwarded-Proto";
+        private string XForwardedForHeaderName
+        {
+            get { return _options.ForwardedForHeaderName; }
+        }
+        private string XForwardedHostHeaderName
+        {
+            get { return _options.ForwardedHostHeaderName; }
+        }
+        private string XForwardedProtoHeaderName
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs]: style: either remove these and access options directly, or at least move them down after the constructor

File: src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersOptions.cs [hunk 2]
```diff
@@ -10,6 +10,38 @@ namespace Microsoft.AspNetCore.Builder
...
+
+        /// <summary>
+        /// X-Forwarded-Proto
+        /// </summary>
+        public const string DefaultForwardedProtoHeaderName = "X-Forwarded-Proto";
+
+        /// <summary>
+        /// Use this header name instead of X-Forwarded-For
+        /// </summary>
+        public string ForwardedForHeaderName { get; set; } = DefaultForwardedForHeaderName;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersOptions.cs]: In the ForwardedHeadersMiddleware constructor validate that these are not null or empty

File: src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersOptions.cs [hunk 3]
```diff
@@ -10,6 +10,38 @@ namespace Microsoft.AspNetCore.Builder
     public class ForwardedHeadersOptions
     {
         /// <summary>
+        /// X-Forwarded-For
+        /// </summary>
+        public const string DefaultForwardedForHeaderName = "X-Forwarded-For";
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersOptions.cs]: We have precedence for separating these to their own class.
https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationDefaults.cs
How about ForwardedHeaderNames.XForwardedFor, etc..

File: src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs (hunk 1 above)
Comment (troydai) [inline src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs]: style: also you can use a more concise form of get-only property
```
private string XForwardedForHeaderName => _options.ForwardedForHeaderName;
```

File: src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersDefaults.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    /// <summary>
+    /// Default values related to <see cref="ForwardedHeadersMiddleware"/> middleware
+    /// </summary>
+    /// <seealso cref="Microsoft.AspNetCore.Builder.ForwardedHeadersOptions"/>
+    public static class ForwardedHeadersDefaults
+    {
+        /// <summary>
+        /// X-Forwarded-For
+        /// </summary>
+        public const string XForwardedForHeaderName = "X-Forwarded-For";
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersDefaults.cs]: Make these `public static readonly`. We avoid `public const` APIs.

**Comment by Tratcher** [comment]
Thanks for being responsive.

**Comment by moander** [comment]
@Tratcher Thanks for reviewing :)

**Pull request merged**